File: faker-egy-names-php/src/Provider.php
<?php

declare(strict_types=1);

namespace Afify\FakerEgyNames;

use Faker\Provider\Base;

/**
 * Egyptian patronymic names via the same generate() rules as faker-egy-names for Python.
 *
 * Snake_case methods match the Python API. CamelCase aliases are for FakerPHP style.
 */
class Provider extends Base
{
    private Engine $engine;

    public function __construct(\Faker\Generator $generator, ?Engine $engine = null)
    {
        parent::__construct($generator);
        $this->engine = $engine ?? new Engine();
    }

    public function egyptian_name(
        ?string $gender = null,
        ?string $religion = null,
        ?int $length = null,
        bool $family_name = true,
        ?string $frequency = null,
        ?int $seed = null,
    ): GeneratedName {
        return $this->generateName($gender, $religion, $length, $family_name, $frequency, $seed);
    }

    public function egyptianName(
        ?string $gender = null,
        ?string $religion = null,
        ?int $length = null,
        bool $familyName = true,
        ?string $frequency = null,
        ?int $seed = null,
    ): GeneratedName {
        return $this->egyptian_name($gender, $religion, $length, $familyName, $frequency, $seed);
    }

    /** @return string|array{0: string, 1: string} */
    public function egyptian_full_name(
        string $lang = 'en',
        ?string $gender = null,
        ?string $religion = null,
        ?int $length = null,
        bool $family_name = true,
        ?string $frequency = null,
        ?int $seed = null,
    ): string|array {
        $name = $this->generateName($gender, $religion, $length, $family_name, $frequency, $seed);

        return self::localized($name->ar, $name->en, $lang);
    }

    /** @return string|array{0: string, 1: string} */
    public function egyptianFullName(
        string $lang = 'en',
        ?string $gender = null,
        ?string $religion = null,
        ?int $length = null,
        bool $familyName = true,
        ?string $frequency = null,
        ?int $seed = null,
    ): string|array {
        return $this->egyptian_full_name($lang, $gender, $religion, $length, $familyName, $frequency, $seed);
    }

    /** @return string|array{0: string, 1: string} */
    public function egyptian_person(
        string $lang = 'en',
        ?string $gender = null,
        ?string $religion = null,
        ?int $length = null,
        bool $family_name = true,
        ?string $frequency = null,
        ?int $seed = null,
    ): string|array {
        return $this->slot(0, $lang, $gender, $religion, $length, $family_name, $frequency, $seed);
    }

    /** @return string|array{0: string, 1: string} */
    public function egyptianPerson(
        string $lang = 'en',
        ?string $gender = null,
        ?string $religion = null,
        ?int $length = null,
        bool $familyName = true,
        ?string $frequency = null,
        ?int $seed = null,
    ): string|array {
        return $this->egyptian_person($lang, $gender, $religion, $length, $familyName, $frequency, $seed);
    }

    /** @return string|array{0: string, 1: string} */
    public function egyptian_father(
        string $lang = 'en',
        ?string $gender = null,
        ?string $religion = null,
        ?int $length = null,
        bool $family_name = true,
        ?string $frequency = null,
        ?int $seed = null,
    ): string|array {
        return $this->slot(1, $lang, $gender, $religion, $length, $family_name, $frequency, $seed);
    }

    /** @return string|array{0: string, 1: string} */
    public function egyptianFather(
        string $lang = 'en',
        ?string $gender = null,
        ?string $religion = null,
        ?int $length = null,
        bool $familyName = true,
        ?string $frequency = null,
        ?int $seed = null,
    ): string|array {
        return $this->egyptian_father($lang, $gender, $religion, $length, $familyName, $frequency, $seed);
    }

    /** @return string|array{0: string, 1: string} */
    public function egyptian_grandfather(
        string $lang = 'en',
        ?string $gender = null,
        ?string $religion = null,
        ?int $length = null,
        bool $family_name = true,
        ?string $frequency = null,
        ?int $seed = null,
    ): string|array {
        return $this->slot(2, $lang, $gender, $religion, $length, $family_name, $frequency, $seed);
    }

    /** @return string|array{0: string, 1: string} */
    public function egyptianGrandfather(
        string $lang = 'en',
        ?string $gender = null,
        ?string $religion = null,
        ?int $length = null,
        bool $familyName = true,
        ?string $frequency = null,
        ?int $seed = null,
    ): string|array {
        return $this->egyptian_grandfather($lang, $gender, $religion, $length, $familyName, $frequency, $seed);
    }

    /** @return string|array{0: string, 1: string} */
    public function egyptian_family(
        string $lang = 'en',
        ?string $gender = null,
        ?string $religion = null,
        ?int $length = null,
        bool $family_name = true,
        ?string $frequency = null,
        ?int $seed = null,
    ): string|array {
        $name = $this->generateName($gender, $religion, $length, $family_name, $frequency, $seed);
        [, , , $family] = self::roles($name, $family_name);

        return self::localized($family[0], $family[1], $lang);
    }

    /** @return string|array{0: string, 1: string} */
    public function egyptianFamily(
        string $lang = 'en',
        ?string $gender = null,
        ?string $religion = null,
        ?int $length = null,
        bool $familyName = true,
        ?string $frequency = null,
        ?int $seed = null,
    ): string|array {
        return $this->egyptian_family($lang, $gender, $religion, $length, $familyName, $frequency, $seed);
    }

    /** @return string|array{0: string, 1: string} */
    private function slot(
        int $index,
        string $lang,
        ?string $gender,
        ?string $religion,
        ?int $length,
        bool $familyName,
        ?string $frequency,
        ?int $seed,
    ): string|array {
        $name = $this->generateName($gender, $religion, $length, $familyName, $frequency, $seed);
        $roles = self::roles($name, $familyName);

        return self::localized($roles[$index][0], $roles[$index][1], $lang);
    }

    private function generateName(
        ?string $gender,
        ?string $religion,
        ?int $length,
        bool $familyName,
        ?string $frequency,
        ?int $seed,
    ): GeneratedName {
        $seed ??= $this->generator->numberBetween(0, 0x7fffffff);

        return $this->engine->generate(
            $gender,
            $religion,
            $length,
            $familyName,
            $frequency,
            $seed,
        );
    }

    /** @return string|array{0: string, 1: string} */
    private static function localized(string $ar, string $en, string $lang): string|array
    {
        return match (strtolower($lang ?: 'en')) {
            'ar' => $ar,
            'both' => [$ar, $en],
            default => $en,
        };
    }

    /**
     * @return array{0: array{0: string, 1: string}, 1: array{0: string, 1: string}, 2: array{0: string, 1: string}, 3: array{0: string, 1: string}}
     */
    private static function roles(GeneratedName $name, bool $familyName): array
    {
        $partsAr = $name->partsAr;
        $partsEn = $name->partsEn;
        $family = ['', ''];
        if ($familyName && count($partsAr) > 1) {
            $family = [array_pop($partsAr), array_pop($partsEn)];
        }

        $slot = static function (int $index) use ($partsAr, $partsEn): array {
            if ($index < count($partsAr)) {
                return [$partsAr[$index], $partsEn[$index]];
            }

            return ['', ''];
        };

        return [$slot(0), $slot(1), $slot(2), $family];
    }
}

class_alias(Provider::class, __NAMESPACE__ . '\\EgyptianNamesProvider');

File: faker-egy-names/src/faker_egy_names/provider.py
"""Faker provider that forwards to egy-names.generate().

No sampling, filtering, or transliteration lives here. Slot strings are
indexes into the chain the engine already returned.
"""

from __future__ import annotations

from typing import Optional, Tuple, Union

from egy_names import EgyNames, GeneratedName
from faker import Faker
from faker.providers import BaseProvider

_engine = EgyNames()

Localized = Union[str, Tuple[str, str]]


def egyptian_faker(*args, **kwargs) -> Faker:
    """Return a Faker instance with this provider already registered."""
    fake = Faker(*args, **kwargs)
    fake.add_provider(Provider)
    return fake


def _localized(ar: str, en: str, lang: str) -> Localized:
    key = (lang or "en").lower()
    if key == "ar":
        return ar
    if key == "both":
        return (ar, en)
    return en


def _roles(
    name: GeneratedName,
    family_name: bool,
) -> Tuple[Tuple[str, str], Tuple[str, str], Tuple[str, str], Tuple[str, str]]:
    """Map generate() parts onto person / father / grandfather / family."""
    parts_ar = list(name.parts_ar)
    parts_en = list(name.parts_en)
    family = ("", "")
    if family_name and len(parts_ar) > 1:
        family = (parts_ar[-1], parts_en[-1])
        parts_ar = parts_ar[:-1]
        parts_en = parts_en[:-1]

    def slot(index: int) -> Tuple[str, str]:
        if index < len(parts_ar):
            return (parts_ar[index], parts_en[index])
        return ("", "")

    return slot(0), slot(1), slot(2), family


class Provider(BaseProvider):
    """Egyptian patronymic names via egy-names 0.3.6."""

    def egyptian_name(
        self,
        gender: Optional[str] = None,
        religion: Optional[str] = None,
        length: Optional[int] = None,
        family_name: bool = True,
        frequency: Optional[str] = None,
        seed: Optional[int] = None,
    ) -> GeneratedName:
        """Return one grounded chain from the engine."""
        return self._generate(
            gender=gender,
            religion=religion,
            length=length,
            family_name=family_name,
            frequency=frequency,
            seed=seed,
        )

    def egyptian_full_name(
        self,
        lang: str = "en",
        *,
        gender: Optional[str] = None,
        religion: Optional[str] = None,
        length: Optional[int] = None,
        family_name: bool = True,
        frequency: Optional[str] = None,
        seed: Optional[int] = None,
    ) -> Localized:
        """Full patronymic name as a string (or `(ar, en)` when `lang='both'`)."""
        name = self._generate(
            gender=gender,
            religion=religion,
            length=length,
            family_name=family_name,
            frequency=frequency,
            seed=seed,
        )
        return _localized(name.ar, name.en, lang)

    def egyptian_person(
        self,
        lang: str = "en",
        *,
        gender: Optional[str] = None,
        religion: Optional[str] = None,
        length: Optional[int] = None,
        family_name: bool = True,
        frequency: Optional[str] = None,
        seed: Optional[int] = None,
    ) -> Localized:
        """Slot 1 — the person's given name."""
        return self._slot(
            0,
            lang,
            gender=gender,
            religion=religion,
            length=length,
            family_name=family_name,
            frequency=frequency,
            seed=seed,
        )

    def egyptian_father(
        self,
        lang: str = "en",
        *,
        gender: Optional[str] = None,
        religion: Optional[str] = None,
        length: Optional[int] = None,
        family_name: bool = True,
        frequency: Optional[str] = None,
        seed: Optional[int] = None,
    ) -> Localized:
        """Slot 2 — father. Empty if the generated chain has no father slot."""
        return self._slot(
            1,
            lang,
            gender=gender,
            religion=religion,
            length=length,
            family_name=family_name,
            frequency=frequency,
            seed=seed,
        )

    def egyptian_grandfather(
        self,
        lang: str = "en",
        *,
        gender: Optional[str] = None,
        religion: Optional[str] = None,
        length: Optional[int] = None,
        family_name: bool = True,
        frequency: Optional[str] = None,
        seed: Optional[int] = None,
    ) -> Localized:
        """Slot 3 — grandfather. Empty if the generated chain is too short."""
        return self._slot(
            2,
            lang,
            gender=gender,
            religion=religion,
            length=length,
            family_name=family_name,
            frequency=frequency,
            seed=seed,
        )

    def egyptian_family(
        self,
        lang: str = "en",
        *,
        gender: Optional[str] = None,
        religion: Optional[str] = None,
        length: Optional[int] = None,
        family_name: bool = True,
        frequency: Optional[str] = None,
        seed: Optional[int] = None,
    ) -> Localized:
        """Final clan / toponymic surname. Empty when `family_name=False`."""
        name = self._generate(
            gender=gender,
            religion=religion,
            length=length,
            family_name=family_name,
            frequency=frequency,
            seed=seed,
        )
        _person, _father, _grandfather, family = _roles(name, family_name)
        return _localized(family[0], family[1], lang)

    def _slot(
        self,
        index: int,
        lang: str,
        *,
        gender: Optional[str],
        religion: Optional[str],
        length: Optional[int],
        family_name: bool,
        frequency: Optional[str],
        seed: Optional[int],
    ) -> Localized:
        name = self._generate(
            gender=gender,
            religion=religion,
            length=length,
            family_name=family_name,
            frequency=frequency,
            seed=seed,
        )
        roles = _roles(name, family_name)
        ar, en = roles[index]
        return _localized(ar, en, lang)

    def _generate(
        self,
        *,
        gender: Optional[str] = None,
        religion: Optional[str] = None,
        length: Optional[int] = None,
        family_name: bool = True,
        frequency: Optional[str] = None,
        seed: Optional[int] = None,
    ) -> GeneratedName:
        if seed is None:
            seed = self.generator.random.getrandbits(32)
        return _engine.generate(
            count=1,
            gender=gender,
            religion=religion,
            length=length,
            family_name=family_name,
            frequency=frequency,
            seed=seed,
        )[0]


EgyptianNamesProvider = Provider
